[Zheng Teck] iP

laughingkid-sg opened the pull request:
# PR Summary 📄
This PR tracks the progress of the iP over time. This greenfield project will be updated over time with additional functionality and turn it into a usable task tracker. The forked and working copy of this version of the project can be found [here](https://github.com/laughingkid-sg/ip).

## Purpose
This PR provides a convenient way for the CS2103T team to access the current state of this iP code from one location.

## PR Revision
This PR Request is updated weekly to reflect the progress:

- [x] Week 1
- [x] Week 2
- [x] Week 3
- [x] Week 4
- [ ] Week 5
- [ ] Week 6

## Features Implemented (By Level)
1. Welcome and closing messages
2. Task creation and listing 
3. Task Completion tracking
4. Task classifation (ToDo, Event, Deadline)
5. Error handling
6. Task deleteion
7. Saving and loading of task(s) from local presistent stroage
8. Date and time formatting for tasks
9. Search of keywords in task

## Main Application
The entry point for the application can be found in the Duke class.
```java
    public static void main(String[] args) {
        new Duke().run();
    }
```

## Benefits of Duke
- Easy to learn and use
- Offline platform
- CLI Support


## Example
> _todo homework_
> ____________________________________________________________
> Got it. I've added this task:
>  _[T][ ] homework_
> Now you have 6 tasks in the list.
> ____________________________________________________________

## Beta Release (v0.1)
A beta release of the application can be accessed here [here](https://github.com/laughingkid-sg/ip/releases/tag/v0.1).

You will need to have java installed.

The user will run the jar file in the following way only:

- Copy the jar file into an empty folder
- Open a command window in that folder
- Run the command java -jar {filename}.jar e.g., `java -jar individual.jar` (i.e., run the command in the same folder as the jar file)

Commit:
Add Gradle support

laughingkid-sg committed:
Update .gitignore

laughingkid-sg committed:
Add Greet Message

laughingkid-sg committed:
Add Echo and Exit Implementation

laughingkid-sg committed:
Add Add and List Implementation

laughingkid-sg committed:
Added Done Implementation

- Enable the user to mark a task as done

laughingkid-sg committed:
Added ToDos, Events, Deadlines Implementation

- Added ToDos, Events, Deadlines classes
- Updated Main to cater to new commands
- Update printing details

laughingkid-sg committed:
Add Handle Errors Implantation

laughingkid-sg committed:
Add Delete Implementation

laughingkid-sg committed:
Fix Minor Bugs

- Add Empty String
- Deadline and event formatting
- Fix while statement

laughingkid-sg committed:
Implemented Testing

laughingkid-sg committed:
Update Error Message

- Additional error messages

laughingkid-sg committed:
Update to Enums

laughingkid-sg committed:
Minor Fix to Function

- Change IsDone to getIsDone()

laughingkid-sg committed:
Merge branch 'week-2-merge'

Fork wrong repo

laughingkid-sg committed:
Add Save and Loading of Tasks

laughingkid-sg committed:
Add parsing for Date and Time Tasks

laughingkid-sg committed:
Merge branch 'branch-Level-8'

laughingkid-sg committed:
Update Constructors

laughingkid-sg committed:
Refactoring OOP

Refactor Duke, Ui, Storage, Parser, TaskList, Command

To refactor the code to extract out closely related code as classes.

Code is abstracted to 5 main classes and 1 abstract class Command

laughingkid-sg committed:
Divide classes into packages

laughingkid-sg committed:
Add JUnit Tests

Add 3 JUnit tests

To demo JUnit test capabilities and perform simple testing.
The tested capabilities are exit parsing, empty storage file and add ToDo to a TaskList.

laughingkid-sg committed:
Create MANIFEST.MF

laughingkid-sg committed:
Add Find Implementation

Give users a way to find a task by searching for a keyword.

laughingkid-sg committed:
Minor Refactoring

laughingkid-sg committed:
Add Cannot Find Message

laughingkid-sg committed:
Reactor to Java coding standard

laughingkid-sg committed:
Add JavaDoc

laughingkid-sg committed:
Merge branch 'branch-A-CodingStandard'

laughingkid-sg committed:
Merge branch 'branch-Level-9'

laughingkid-sg committed:
Update README.md

Fixed auto rename of the project.

BrendonLau commented on src/main/java/duke/io/Storage.java:
```diff
@@ -0,0 +1,103 @@
+            LocalDate date;
+            LocalTime time;
+            while (s.hasNext()) {
+                String[] inputLine = s.nextLine().split("\\|");
+                String type = inputLine[0];
+                boolean isDone = inputLine[1].equals("1");
+                switch (type) {
+                    case "T":
+                        taskList.addToDo(inputLine[2],isDone);
+                        break;
+                    case "E":
+                        input = inputLine[3].split(" ");
+                        if (input.length != 2) {
+                            throw new DukeException(Ui.MSG_FILEREADERROR);
+                        }
+                        date = Parser.formatDate(input[0]);
+                        time = Parser.formatTime(date, input[1]);
+                        taskList.addEvent(inputLine[2],isDone, date, time);
+                        break;
+                    case "D":
+                        input = inputLine[3].split(" ");
+                        if (input.length != 2) {
+                            throw new DukeException(Ui.MSG_FILEREADERROR);
+                        }
+                        date = Parser.formatDate(input[0]);
+                        time = Parser.formatTime(date, input[1]);
+                        taskList.addDeadline(inputLine[2],date,time);
+                }
```
Maybe to improve readability is to assign variables to inputLine[0/1/2/3]. I am guilty of this as well.

laughingkid-sg replied:
Hi, I do not understand what do you mean by `assign variables to inputLine[0/1/2/3]`, could you please explain further. Cheers!

BrendonLau replied:
I was confused in a way I have to make sense of what is input[3] or input[0] in different cases. Upon closer look, you did specify what is inputLine[0] by assigning it to type and date = input[0] etc. I think it is fine. Maybe not use reassign input but use dateTime to make it more readable?

BrendonLau reviewed (commented):
Overall, well done on the abstraction and LGTM!

aweijun commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,170 @@
+            case DELETE:
+                return formatCmdWithIdSelection(action, inputTxt);
+            default:
+                throw new DukeException(Ui.MSG_INVALIDCMD);
```
Really neat code!

Nice use of switch cases, to simplify to code.

aweijun commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,170 @@
+    }
+
+    private Command formatCmdWithTime(CommandType commandType, String inputTxt) throws DukeException {
+        String[] formatInputTxt, dateTime;
```
Maybe the variable names could be plural instead to indicate that this is a list?

aweijun commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,170 @@
+import duke.command.ExitCommand;
+import duke.command.FindCommand;
+import duke.command.InvalidCommand;
+import duke.command.ListCommand;
```
Great job explicitly importing all the imported class!

aweijun commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,73 @@
+    public static final String MSG_TASKALREADYDONE = "☹ OOPS!!! task.Task already completed!";
+
+    public static final String MSG_EMPTYTASK = "☹ OOPS!!! No task exists!";
+
+    public static final String MSG_FILEREADERROR = "☹ OOPS!!! There was an error loading the file.";
+
+    public static final String MSG_INVALIDDATETIMEFORMAT = "☹ OOPS!!! Wrong format for Date or Time. " + "Please use [dd-MM-yyyy HH:mm] format. i.e. 30-05-1959 13:59";
+
+    public static final String MSG_INVALIDDATETIME = "☹ OOPS!!! You can't set a task in the past!";
+
+    public static final String MSG_FILEWRITEERROR = "☹ OOPS!!! We can't save your records! This is likely due to write permissions on your system.";
+
+    public static String taskListMsg(TaskList taskList) {
+        ArrayList<Task> tasks = taskList.getTaskList();
+        StringBuffer result = new StringBuffer("Here are the tasks in your list: \n");
+        for (int i = 1; i <= tasks.size(); i++) {
+            result.append(String.format("%d.%s", i, tasks.get(i - 1)));
+            if (i != tasks.size()) {
+                result.append("\n");
+            }
+        }
+        return result.toString();
+    }
+
+    public static String completeTaskMsg(String description) {
+        return "Nice! I've marked this task as done: \n  " + description;
+    }
+
+    public static String deleteTaskMsg(String description, int totalTask) {
+        return String.format("Noted. I've removed this task: \n  %s\nNow you have %d tasks in the list.", description, totalTask);
+    }
+
+    public static String addTaskMsg(String description, int totalTask) {
+        return String.format("Got it. I've added this task:\n  %s\nNow you have %d tasks in the list.", description, totalTask);
+    }
+
+    public static void print(String input) {
+        System.out.println(MSG_LINE_SEPARATOR + input + "\n" + MSG_LINE_SEPARATOR);
+    }
+}
```
There seems to be missing java documentation in this file.

I understand that it is not a requirement, just highlighting it in case this was unintentional given how the rest of the files have java documentation

aweijun commented on src/main/java/duke/io/Storage.java:
```diff
@@ -0,0 +1,103 @@
+                    case "E":
+                        input = inputLine[3].split(" ");
+                        if (input.length != 2) {
+                            throw new DukeException(Ui.MSG_FILEREADERROR);
+                        }
+                        date = Parser.formatDate(input[0]);
+                        time = Parser.formatTime(date, input[1]);
+                        taskList.addEvent(inputLine[2],isDone, date, time);
+                        break;
+                    case "D":
+                        input = inputLine[3].split(" ");
+                        if (input.length != 2) {
+                            throw new DukeException(Ui.MSG_FILEREADERROR);
+                        }
```
Good job trying to catch other possible run time errors!

aweijun reviewed (commented):
Great work! 
Couldn't find any major violations or examples of bad code quality, only few minor suggestions

I learned a lot from reviewing your code.

laughingkid-sg committed:
Fixed Minor Bugs

Adjusted access modifier for methods that are only access by the class
itself.

Updated parser to replace all "|" instance to "-" to ensure it does not
affect saving to local persistent storage.

Added missing constants and methods for Find implementation due to
merge.

Fixed Index Out of Bounds for Done command due to wrong order of if
statement to check out of range item.

laughingkid-sg committed:
Update Javadoc

Add Javadoc for Find, Duke and Ui implementation.

laughingkid-sg committed:
Add Help Implementation

laughingkid-sg committed:
Add Undo Implementation

The Undo implementation allows the user to mark a task as incomplete.

laughingkid-sg committed:
Merge branch 'add-gradle-support' into branch-A-Gradle

laughingkid-sg committed:
Add CheckStyle

laughingkid-sg committed:
Update styling for String Block

laughingkid-sg committed:
Merge branch 'branch-A-CheckStyle'

laughingkid-sg committed:
Update build.gradle

laughingkid-sg committed:
Merge branch 'branch-A-Gradle'

laughingkid-sg committed:
Update General CheckStyle

laughingkid-sg committed:
Add JavaFx

laughingkid-sg committed:
Update Error Handling for GUI

laughingkid-sg committed:
Update GUI

laughingkid-sg committed:
Add Assertions

laughingkid-sg committed:
Improve Code Quality

Improve code overall quality by applying the recommended topics

Removed Deep Nesting
Avoid Magic Numbers
Make the Code Obvious
Remove Unnecessary Imports

laughingkid-sg committed:
Merge pull request #2 from laughingkid-sg/branch-A-Assertions

Add Assertions

laughingkid-sg committed:
Merge branch 'master' into branch-A-CodeQuality

laughingkid-sg committed:
Merge pull request #3 from laughingkid-sg/branch-A-CodeQuality

Improve Code Quality

laughingkid-sg committed:
Update Task Types

Storage contains the types (declared as constant Strings)
of the task for importing declared independently from the Task classes.

The common value of the types across the Task classes and Storage class
cause code duplication.

The constants are declare in the sub-classes of the Task classes and
the getType() method is made public so that the storage class can access
the values.

This is to ensure that the types used in Storage during import is
consistent with the types used during export.

laughingkid-sg committed:
Add Reminders Implementation

Currently there is no way for the users to get a list of task that is
due within a specific period of time.

To improve the overall user experience, the reminders command is
introduced.

The reminders command allow the user to get deadlines that are within
certain days.

This allows the user to look ahead the the deadlines for next 7 days
using the reminders command (i.e. reminders 7).

Usage of command has also been added to the help command list.

laughingkid-sg committed:
Update Ui Package to Lowercase

laughingkid-sg committed:
Update GUI

- Updated color to 4 different style - User Command, Bot Reply,
  User Command (Invalid), Bot Reply (Invalid)
- Improve overall Ui to give it a more pleasing look
- Added colors to invalid commands
- Update profile pictures

laughingkid-sg committed:
Update Launcher.java

laughingkid-sg committed:
Create Ui.png

laughingkid-sg committed:
Update README.md

laughingkid-sg committed:
Merge branch 'master' of https://github.com/laughingkid-sg/ip

laughingkid-sg committed:
Update build.gradle

laughingkid-sg committed:
Update Content

laughingkid-sg committed:
Update README.md

laughingkid-sg committed:
Update reminders to exclude completed tasks

laughingkid-sg committed:
Merge branch 'master' of https://github.com/laughingkid-sg/ip
